<a href="https://colab.research.google.com/github/sanghaimuskan/Breast-Cancer-Detection-using-KNN/blob/master/Breast_Cancer_using_Decision_tree.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import files
uploaded = files.upload()

Saving Breast_cancer.csv to Breast_cancer.csv


In [7]:
data = pd.read_csv('Breast_cancer.csv')
data.head()

id diagnosis  ...  fractal_dimension_worst  Unnamed: 32
0    842302         M  ...                  0.11890          NaN
1    842517         M  ...                  0.08902          NaN
2  84300903         M  ...                  0.08758          NaN
3  84348301         M  ...                  0.17300          NaN
4  84358402         M  ...                  0.07678          NaN

[5 rows x 33 columns]

# *Remove id and unnamed: 32 column from the dataset* 

In [8]:
data = data.drop(['id','Unnamed: 32'], axis = 1)
data.shape

(569, 31)

In [9]:
data.head()

diagnosis  radius_mean  ...  symmetry_worst  fractal_dimension_worst
0         M        17.99  ...          0.4601                  0.11890
1         M        20.57  ...          0.2750                  0.08902
2         M        19.69  ...          0.3613                  0.08758
3         M        11.42  ...          0.6638                  0.17300
4         M        20.29  ...          0.2364                  0.07678

[5 rows x 31 columns]

# **Convert M and B into 1 and 0 respectedly**

In [10]:
def diagnosis_value(diagnosis):
  if diagnosis == 'M':
    return 1
  else:
    return 0

data['diagnosis'] = data['diagnosis'].apply(diagnosis_value)

In [74]:
X = np.array(data.iloc[:, 1:])
y = np.array(data['diagnosis'])

In [75]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 42)

# **We will be using Decision Tree with max_depth=5**

In [76]:
from sklearn.tree import DecisionTreeClassifier
D_Tree = DecisionTreeClassifier( max_depth = 5)

In [77]:
D_Tree.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=5, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [78]:
D_Tree.score(X_test, y_test)

0.9415204678362573

## ***The accuracy of the model is 94.15%***

## *We will be using another metric called confusion matrix to check which are predicted wrong*

In [79]:
from sklearn.model_selection import cross_val_predict
pred = cross_val_predict(D_Tree , X_train, y_train, cv = 3)

In [80]:
print(pred.shape)

print(y_train.shape)

(398,)
(398,)


In [81]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_train, pred)

array([[235,  14],
       [ 13, 136]])